enhancements/sig-architecture/224-event-based-manifestwork/README.md

```diff
@@ -0,0 +1,1042 @@
+for `ManifestWork` with MQTT options, the builder returns a work client holder, we can get the work client set from this
+holder, e.g.
+
+```golang
+workClientHolder, err := workclient.NewWorkClientBuilder("work-agent-id").
+		WithHubKubeConfig(hubKubeConfig).
+		WithMQTTOptions(o.mqttOptions).
+		WithClusterName(o.SpokeClusterName).
+		WithCloudEventTypes(workclient.ManifestBundle, workclient.Manifest).
+		NewWorkClientHolder(ctx)
+if err != nil {
+		return err
+}
+
+workClientSet := hubWorkClientHolder.GetWorkClientSet()
+```
+
+Using the work client set, developers can access the ManifestWork or create a Kubernetes informer and lister. This
+process closely mirrors what developers typically do in a Kubernetes environment.
+
+```golang
+workInformerFactory := workinformers.NewSharedInformerFactoryWithOptions(workClientSet, 5*time.Minute, workinformers.WithNamespace(o.SpokeClusterName))
+manifestWorkClient := workClientSet.WorkV1().ManifestWorks(o.SpokeClusterName)
+manifestWorkInformer := workInformerFactory.Work().V1().ManifestWorks()
+manifestWorkLister := workInformerFactory.Work().V1().ManifestWorks().Lister().ManifestWorks(o.SpokeClusterName)
+```
+
+### Failure Handling
+
+To ensure the integrity of the event-based manifestworks system during source or work agent restart or reconnection, the following optimized resync mechanism is employed:
+
+1. On start, the source subscribes to the spec resync event topic `/sources/clusters/+/specresync` from the broker.
+2. On start, the work agent subscribes to the status resync event topic `/sources/+/clusters/statusresync` from the broker.
+3. Upon work agent restarts or reconnects, the work agent generates a spec resync event with resource versions and sends it to the broker.
+4. Upon source restarts or reconnects, the source calculates a status hash of its owned resources and sends a status resync event with the resource hash to the broker.
+5. The source compares the resource versions in the spec resync event message with the hub's state and generates spec events to the broker for the work agent to receive missing spec events.
+6. The work agent compares the status hashes in the status resync event message with its own state and generates status events to the broker for the source to receive missing status events.
+
+### Impact to Relevant Components
+
+Using cloud event based manifestwork can streamline resource delivery and status retrieval for various components managing resources. For instance, the policy controller can utilize this feature, eliminating the necessity for Policy agent to report status to the hub cluster namespace. Instead, Work-agent will handle reporting the policy status. The Policy propagator controller will aggregate the status through status report cloud event, obviating the need for replicatedPolicy. Additionally, there will be no cluster namespace in the hub side. This approach enhances efficiency and simplifies resource management across managed clusters.
```

Comment:
Is the idea that ManifestWork would deliver the replicated policy on the managed cluster side?

We need a local cache on the managed cluster side to account for when the cluster is disconnected from the Hub.

Reply:
yes, there is a cache in managed cluster side for the ManifestWorks from the broker.

Reply:
yes, I think we can deliver the replicated policy with the manifestwork

Reply:
For work agent, it uses the memory as the cache, after the pod restarts, the work agent will send a resync request to the source to sync the resources it maintained  and rebuild its cache

Reply:
The cache, based on [client-go's store](https://pkg.go.dev/k8s.io/client-go/tools/cache#Store) implementation, automatically clears upon pod restarts.